### PR TITLE
Pod security policies

### PR DESCRIPTION
**Pull request by discreet**
The below PR has been open since September 20, 2019. The author has life events going on that do not allow them to finish the work. This PR is built off of their fork and branch and makes an attempt to resolve the blocking merge conflicts.

https://github.com/terraform-providers/terraform-provider-kubernetes/pull/624

### CODE REVIEW

File: website/docs/r/pod_security_policy.html.markdown [hunk 1]
```diff
@@ -0,0 +1,179 @@
+---
+layout: "kubernetes"
+page_title: "Kubernetes: kubernetes_pod_security_policy"
+sidebar_current: "docs-kubernetes-resource-pod-security-policy"
+description: |-
+  A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod specification.
+---
+
+# kubernetes_api_service
```
Comment (jrhouston) [inline website/docs/r/pod_security_policy.html.markdown:9]: This should say `kubernetes_pod_security_policy`

File: kubernetes/provider.go [hunk 2]
```diff
@@ -148,36 +148,35 @@ func Provider() terraform.ResourceProvider {
...
+			"kubernetes_priority_class":            resourceKubernetesPriorityClass(),
+			"kubernetes_replication_controller":    resourceKubernetesReplicationController(),
+			"kubernetes_role_binding":              resourceKubernetesRoleBinding(),
+			"kubernetes_resource_quota":            resourceKubernetesResourceQuota(),
+			"kubernetes_role":                      resourceKubernetesRole(),
+			"kubernetes_secret":                    resourceKubernetesSecret(),
+			"kubernetes_service":                   resourceKubernetesService(),
+			"kubernetes_service_account":           resourceKubernetesServiceAccount(),
+			"kubernetes_stateful_set":              resourceKubernetesStatefulSet(),
+			"kubernetes_storage_class":             resourceKubernetesStorageClass(),
```
Comment (jrhouston) [inline kubernetes/provider.go:179]: `kubernetes_validating_webhook_configuration` and `kubernetes_mutating_webhook_configuration` have gone missing here!

File: kubernetes/structure_pod_security_policy_spec.go [hunk 3]
```diff
@@ -0,0 +1,658 @@
+package kubernetes
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+
+	v1 "k8s.io/api/core/v1"
+	v1beta1 "k8s.io/api/extensions/v1beta1"
```
Comment (jrhouston) [inline kubernetes/structure_pod_security_policy_spec.go:9]: For PodSecurityPolicy we want to use `"k8s.io/api/policy/v1beta1"` as it has been removed from `extensions` in kubernetes v1.18.

File: website/docs/r/pod_security_policy.html.markdown [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+* `path_prefix` - (Required) the path prefix that the host volume must match. It does not support `*`. Trailing slashes are trimmed when validating the path prefix with a host path. Examples: `/foo` would allow `/foo`, `/foo/` and `/foo/bar`. `/foo` would not allow `/food` or `/etc/foo`
+* `read_only` - (Optional) when set to true, will allow host volumes matching the pathPrefix only if all volume mounts are readOnly.
+
+### `fs_group`
+
+#### Arguments
+
+* `rule` - (Required) the strategy that will dictate what FSGroup is used in the SecurityContext.
+* [`ranges`](#ranges) - (Optional) the allowed ranges of fs groups.  If you would like to force a single fs group then supply a single range with the same start and end. Required for MustRunAs.
```
Comment (jrhouston) [inline website/docs/r/pod_security_policy.html.markdown:137]: This is called `range` in the schema.

File: website/docs/r/pod_security_policy.html.markdown (hunk 1 above)
Comment (discreet) [reply website/docs/r/pod_security_policy.html.markdown]: Fixed

File: kubernetes/structure_pod_security_policy_spec.go (hunk 3 above)
Comment (discreet) [reply kubernetes/structure_pod_security_policy_spec.go]: Fix

File: kubernetes/provider.go (hunk 2 above)
Comment (discreet) [reply kubernetes/provider.go]: Fixed

File: website/docs/r/pod_security_policy.html.markdown (hunk 4 above)
Comment (discreet) [reply website/docs/r/pod_security_policy.html.markdown]: Fixed

File: website/docs/r/pod_security_policy.html.markdown [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+* `rule` - (Required) the strategy that will dictate the allowable RunAsUser values that may be set.
+* [`ranges`](#ranges) - (Optional) the allowed ranges of uids that may be used. If you would like to force a single uid then supply a single range with the same start and end. Required for MustRunAs.
+
+### `run_as_group`
+
+#### Arguments
+
+* `rule` - (Required) the strategy that will dictate the allowable RunAsGroup values that may be set.
+* [`ranges`](#ranges) - (Optional) the allowed ranges of gids that may be used. If you would like to force a single gid then supply a single range with the same start and end. Required for MustRunAs.
```
Comment (jrhouston) [inline website/docs/r/pod_security_policy.html.markdown:151]: This should be `range` too

File: website/docs/r/pod_security_policy.html.markdown [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+#### Arguments
+* `rule` - (Required) the strategy that will dictate the allowable labels that may be set.
+* `se_linux_options` - (Optional) required to run as; required for MustRunAs. More info: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
+
+### `supplemental_groups`
+
+#### Arguments
+
+* `rule` - (Required) the strategy that will dictate what supplemental groups is used in the SecurityContext.
+* [`ranges`](#ranges) - (Optional) the allowed ranges of supplemental groups.  If you would like to force a single supplemental group then supply a single range with the same start and end. Required for MustRunAs.
```
Comment (jrhouston) [inline website/docs/r/pod_security_policy.html.markdown:164]: This one too

File: website/docs/r/pod_security_policy.html.markdown [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+
+### `ranges`
```
Comment (jrhouston) [inline website/docs/r/pod_security_policy.html.markdown:167]: And this one 😁

File: website/docs/r/pod_security_policy.html.markdown (hunk 5 above)
Comment (discreet) [reply website/docs/r/pod_security_policy.html.markdown]: Fixed

File: website/docs/r/pod_security_policy.html.markdown (hunk 6 above)
Comment (discreet) [reply website/docs/r/pod_security_policy.html.markdown]: Fixed

File: website/docs/r/pod_security_policy.html.markdown (hunk 7 above)
Comment (discreet) [reply website/docs/r/pod_security_policy.html.markdown]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy.go [hunk 8]
```diff
@@ -0,0 +1,509 @@
+package kubernetes
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	api "k8s.io/api/extensions/v1beta1"
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy.go:8]: Still using `extensions` here, this needs to be updated to be 

`policy "k8s.io/api/policy/v1beta1"` to match your structured_pod_security_policy.go file

File: kubernetes/resource_kubernetes_pod_security_policy_test.go [hunk 9]
```diff
@@ -0,0 +1,398 @@
+package kubernetes
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	api "k8s.io/api/extensions/v1beta1"
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy_test.go:10]: Same here

File: kubernetes/resource_kubernetes_pod_security_policy.go (hunk 8 above)
Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy.go]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy_test.go (hunk 9 above)
Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy_test.go]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy_test.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+	}
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "kubernetes_pod_security_policy" {
+			continue
+		}
+
+		name := rs.Primary.ID
+
+		resp, err := conn.ExtensionsV1beta1().PodSecurityPolicies().Get(name, meta_v1.GetOptions{})
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy_test.go:196]:
```suggestion
		resp, err := conn.PolicyV1beta1().PodSecurityPolicies().Get(name, meta_v1.GetOptions{})
```

File: kubernetes/resource_kubernetes_pod_security_policy.go [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+			return err
+		}
+		ops = append(ops, *diffOps...)
+	}
+	data, err := ops.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("Failed to marshal update operations: %s", err)
+	}
+	log.Printf("[INFO] Updating PodSecurityPolicy %q: %v", name, string(data))
+	out, err := conn.ExtensionsV1beta1().PodSecurityPolicies().Patch(name, pkgApi.JSONPatchType, data)
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy.go:449]:
```suggestion
	out, err := conn.PolicyV1beta1().PodSecurityPolicies().Patch(name, pkgApi.JSONPatchType, data)
```

File: kubernetes/resource_kubernetes_pod_security_policy.go [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+func resourceKubernetesPodSecurityPolicyDelete(d *schema.ResourceData, meta interface{}) error {
+	conn, err := meta.(KubeClientsets).MainClientset()
+	if err != nil {
+		return err
+	}
+
+	name := d.Id()
+
+	log.Printf("[INFO] Deleting PodSecurityPolicy: %#v", name)
+	err = conn.ExtensionsV1beta1().PodSecurityPolicies().Delete(name, &meta_v1.DeleteOptions{})
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy.go:468]:
```suggestion
	err = conn.PolicyV1beta1().PodSecurityPolicies().Delete(name, &meta_v1.DeleteOptions{})
```

File: kubernetes/resource_kubernetes_pod_security_policy.go [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+func resourceKubernetesPodSecurityPolicyExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	conn, err := meta.(KubeClientsets).MainClientset()
+	if err != nil {
+		return false, err
+	}
+
+	name := d.Id()
+
+	log.Printf("[INFO] Checking PodSecurityPolicy %s", name)
+	_, err = conn.ExtensionsV1beta1().PodSecurityPolicies().Get(name, meta_v1.GetOptions{})
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_pod_security_policy.go:486]:
```suggestion
	_, err = conn.PolicyV1beta1().PodSecurityPolicies().Get(name, meta_v1.GetOptions{})
```

Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy.go]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy.go (hunk 12 above)
Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy.go]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy.go (hunk 11 above)
Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy.go]: Fixed

File: kubernetes/resource_kubernetes_pod_security_policy_test.go (hunk 10 above)
Comment (discreet) [reply kubernetes/resource_kubernetes_pod_security_policy_test.go]: Fixed

**Comment by discreet** [comment]
Wooo glad this is finally ready. Really appreciate your patience with me @jrhouston. This is going to be a very helpful and very needed feature for me to better manage my company's EKS clusters.

**Pull request merged**